lib/Sema/CSDiagnostics.cpp

```diff
@@ -3493,3 +3493,62 @@ bool MutatingMemberRefOnImmutableBase::diagnoseAsError() {
                             diagIDmember);
   return failure.diagnoseAsError();
 }
+
+bool InvalidTupleSplatWithSingleParameterFailure::diagnoseAsError() {
+  auto *anchor = getRawAnchor();
+
+  auto selectedOverload = getChoiceFor(anchor);
+  if (!selectedOverload || !selectedOverload->choice.isDecl())
+    return false;
+
+  auto *choice = selectedOverload->choice.getDecl();
+
+  auto *argExpr = getArgumentExprFor(anchor);
+  if (!argExpr)
+    return false;
+
+  using Substitution = std::pair<GenericTypeParamType *, Type>;
+  llvm::SetVector<Substitution> substitutions;
+  auto paramTy = ParamType.transform([&](Type type) -> Type {
+    if (auto *typeVar = type->getAs<TypeVariableType>()) {
+      type = resolveType(typeVar);
+
+      if (auto *GP = typeVar->getImpl().getGenericParameter()) {
+        substitutions.insert(std::make_pair(GP, type));
+        return GP;
+      }
+    }
+
+    return type;
+  });
+
+  DeclBaseName name = choice->getBaseName();
+
+  std::string subsStr;
+  if (!substitutions.empty()) {
+    subsStr += " [where ";
+    interleave(
+        substitutions,
+        [&subsStr](const Substitution &substitution) {
+          subsStr += substitution.first->getString();
+          subsStr += " = ";
+          subsStr += substitution.second->getString();
+        },
+        [&subsStr] { subsStr += ", "; });
+    subsStr += ']';
+  }
+
+  auto diagnostic =
+      name.isSpecial()
+          ? emitDiagnostic(argExpr->getLoc(),
+                           diag::single_tuple_parameter_mismatch_special,
```

Comment:
Why do we suppress the name for special names? It would be fine to print `subscript` or `init` here, for example.

Reply:
That's how original diagnostic has been setup in the past, I can use `DeclName` to unify.

Reply:
As far as I remember the reason why we did this originally is because most cases there is going to be no label for subscript or initializer with a single parameter so printing it like `subscript 'subscript(_:)'` doesn't really buy us much...

Reply:
Oh, I'd forgotten about the `subscript 'subscript(_:)'` issue. It's fine to leave it split into two diagnostics for this reason.